plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcJoinPushdownSupportModule.java

```diff
@@ -31,7 +31,5 @@ protected void setup(Binder binder)
     {
         configBinder(binder).bindConfig(JdbcJoinPushdownConfig.class);
         bindSessionPropertiesProvider(binder, JdbcJoinPushdownSessionProperties.class);
-
-        configBinder(binder).bindConfigDefaults(JdbcMetadataConfig.class, config -> config.setJoinPushdownEnabled(true));
```

Comment:
JdbcJoinPushdownSupportModule is used only by connectors which support cost-based join pushdown.

Reply:
`JdbcMetadataConfig` your comment:
```
Join pushdown is disabled by default as this is the safer option.
```
And it is enabled by default.

Reply:
I'm assuming that meant that it was unsafe to enable in all cases but is safe if we do it based on the cost based estimate.
I think the cost based estimate doesn't account for reduction in probe output due to dynamic filtering. It might be possible to estimate this based on NDV of the columns being joined. @Dith3r maybe we can try building that into the cost estimation.
Otherwise, I don't think we should outright disable this as this covers a wide variety of data sources and we can't prove that disabling pushdown is the best strategy overall.

Reply:
> I think the cost based estimate doesn't account for reduction in probe output due to dynamic filtering. 

I think we need to consider different join types under assumption that we want to minimize data size fetched from connector

1). Filtering join (e.g. fact table join with dimension table), non-expanding

Filtering join are most common and should always be beneficial to pushdown. Even with perfect DF from Trino, join output will always be smaller than probe and build side

2). Expanding join (multiplies output rows for every probe input row), no filtering

Here DFs are irrelevant (no filtering). Might still be beneficial to pushdown depending on build side size. However, expanding joins are usually hard to estimate.

3). Expanding join that does some filtering also

Maybe we could estimate % of rows filtered and expanding factor for remaining rows? Generally, I would probably avoid pushdown of expanding joins because of uncertainity of stats. Assuming DFs work, then they should do good join at filtering probe side data so we should only get rows that get expanded. Pushing expanding join to connector might explode data produced by connector.

Reply:
> I'm assuming that meant that it was unsafe to enable in all cases but is safe if we do it based on the cost based estimate.

yes

Reply:
> Filtering join are most common and should always be beneficial to pushdown. Even with perfect DF from Trino, join output will always be smaller than probe and build side

Perhaps we should also think about input to the join instead of just join output. All else being equal, the one with smaller input to the join should be able to process it faster. I think in the current estimate we assume that the cost of doing the join is the same in both systems, but that would change if Trino join benefits a lot from reduced input by dynamic filtering while the remote DB doesn't do that optimization.

Reply:
💯  @raunaqmorarka , the cost-based join decision algorithm remains in its v1 form and there definitely is a lot room for improvement.

Reply:
What I'm essentially saying is that we probably could simplify the cost decision to expanding vs non-expanding join. Probably the same could apply to other pushdown. However, by not-pushing join we might prevent further pushdowns (e.g. aggregaton) that could finally reduce produced data to minimum.

This is where exhaustive optimizer would come handy (at least via alternatives in source stage).

> I think in the current estimate we assume that the cost of doing the join is the same in both systems, but that would change if Trino join benefits a lot from reduced input by dynamic filtering while the remote DB doesn't do that optimization.

I would actually assume that by default the cost of doing join in connector is less than cost in Trino, because there are less layers and data transfers in between and we don't actually have bigger pipe to data than remote DB itself. Do we know of databaes which would have worse join implementation even considering data transfers?

Reply:
> What I'm essentially saying is that we probably could simplify the cost decision to expanding vs non-expanding join.

I believe that's the current logic as implemented.
https://github.com/trinodb/trino/blob/095d35adac6697ebdf5ed17b710e4e8a3eb27265/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcJoinPushdownConfig.java#L26-L28

> However, by not-pushing join we might prevent further pushdowns

true

> Hence it's not entirely connector decision do do or not do pushdown

that would require exploratory optimizer

Reply:
> that would require exploratory optimizer

yes, but not full exploratory optimizer, but it would be sufficient to have bunch of alternatives in source stage cc. @assaf2

Reply:
For SQL databases, we should only consider push down if a covering index is present and for Postgres if any of the joined tables is not partitioned, as this case underperforms.

Reply:
> For SQL databases, we should only consider push down if a covering index is present and for Postgres if any of the joined tables is not partitioned, as this case underperforms.

I think it's something we should put into documentation (cc @mosabua ) that users should add indexes on columns they plan to use in joins

Reply:
If this is indeed the case, then first and foremost the join pushdown code should check for necessary indexes being present.